### PR TITLE
Data-base-connection

### PR DESCRIPTION
**Pull request by gandhinikhil**
Add new Feature and more functionality.

### CODE REVIEW

File: src/main/java/com/example/cabbooking/entity/Driver.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package com.example.cabbooking.entity;
+
+import jakarta.persistence.*;
+import lombok.*;
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Driver.java:3]: use deep import

File: src/main/java/com/example/cabbooking/entity/Driver.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    private String name;
+    private String gender;
+    private int age;
+    private boolean isAvailable;
+    @OneToOne(cascade = CascadeType.ALL)
+    @JoinColumn(name = "vehicle_id", referencedColumnName = "id")
+    private Vehicle vehicle;
+    @OneToOne(cascade = CascadeType.ALL)
+    @JoinColumn(name = "location_id", referencedColumnName = "id")
+    private Location loaction;
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Driver.java]: typo

File: src/main/java/com/example/cabbooking/entity/Driver.java (hunk 1 above)
Comment (gandhinikhil) [reply src/main/java/com/example/cabbooking/entity/Driver.java:3]: yes I will cover this

File: src/main/java/com/example/cabbooking/entity/Driver.java (hunk 2 above)
Comment (gandhinikhil) [reply src/main/java/com/example/cabbooking/entity/Driver.java]: yes I have update but yes i will commit again

File: src/main/java/com/example/cabbooking/entity/Driver.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    @OneToMany(mappedBy = "driver", cascade = CascadeType.ALL)
+    private List<Ride> ride;
+    private LocalDateTime localDateTime = LocalDateTime.now();
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Driver.java]: @PrePersist
protected void onCreate() {
    createdAt = LocalDateTime.now();
}
maybe you can check if you can use this ?

File: src/main/java/com/example/cabbooking/entity/Driver.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+
+}
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Driver.java:42]: remove EOF extra lines

File: src/main/java/com/example/cabbooking/entity/Location.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.example.cabbooking.entity;
+
+import jakarta.persistence.*;
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Location.java:2]: deep import

File: src/main/java/com/example/cabbooking/entity/Location.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+@AllArgsConstructor
+@Entity
+@Table(name="locations")
+public class Location {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private long id;
+    private int xDistance;
+    private int yDistance;
+    private LocalDateTime localDateTime = LocalDateTime.now();
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Location.java]: same comment here for localDateTime

File: src/main/java/com/example/cabbooking/entity/Ride.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package com.example.cabbooking.entity;
+
+import jakarta.persistence.*;
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/entity/Ride.java:2]: same

File: src/main/java/com/example/cabbooking/entity/Driver.java (hunk 3 above)
Comment (gandhinikhil) [reply src/main/java/com/example/cabbooking/entity/Driver.java]: yes will check and update

File: src/main/java/com/example/cabbooking/servicetest/DriverService.java [hunk 8]
```diff
@@ -27,19 +42,45 @@ public void addDriver(DriverDetailsVO driverDetailsVO) {
         LocationDetailsVO location = new LocationDetailsVO(locationDetailsVO.getXDistance(), locationDetailsVO.getYDistance());
         DriverDetailsVO driver = new DriverDetailsVO(driverDetailsVO.getName(), driverDetailsVO.getGender(), driverDetailsVO.getAge(), vehicle, location, true);
         drivers.put(driver.getName(), driver);
-        System.out.println(driver.getName()+" "+driver.getGender()+" "+driver.getAge()+" "+driver.isAvailable()+" "+driver.getLocation()+" "+driver.getVehicle());
+        System.out.println(driver.getName() + " " + driver.getGender() + " " + driver.getAge() + " " + driver.isAvailable() + " " + driver.getLocation() + " " + driver.getVehicle());
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/servicetest/DriverService.java]: remove this after debugging is done or comment out

File: src/main/java/com/example/cabbooking/servicetest/DriverService.java [hunk 9]
```diff
@@ -27,19 +42,45 @@ public void addDriver(DriverDetailsVO driverDetailsVO) {
...
-    public List<RideDetailsVO> findAvailableDrivers(LocationDetailsVO source,LocationDetailsVO destination, int maxDistance) {
+    public List<RideDetailsVO> findAvailableDrivers(LocationDetailsVO source, LocationDetailsVO destination, int maxDistance) {
         List<RideDetailsVO> availableDrivers = new ArrayList<>();
         for (DriverDetailsVO driver : drivers.values()) {
-            if (driver.isAvailable() && driver.getLocation().distanceTo(source)<= maxDistance) {
-                double fare = rideService.calculateFare(source,destination);
-                availableDrivers.add(new RideDetailsVO(driver,Math.round(fare*100.0)/100.0));
+            if (driver.isAvailable() && driver.getLocation().distanceTo(source) <= maxDistance) {
+                double fare = rideService.calculateFare(source, destination);
+                availableDrivers.add(new RideDetailsVO(driver, Math.round(fare * 100.0) / 100.0));
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/servicetest/DriverService.java:79]: can we seperate the logic here ? maybe 2 different methods ? one for finding the driver based on whatever business logic you have and another for calculating fare / price of trip , so the code will follow maintainability , easy to add more logic in future

**Review by sanskrati3 — APPROVED** [review]
Rest LGTM 
image

**Comment by gandhinikhil** [comment]
All comment done with required changes

**Pull request merged**